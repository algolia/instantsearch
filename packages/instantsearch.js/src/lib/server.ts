import { walkIndex } from './utils';

import type {
  IndexWidget,
  InitialResults,
  InstantSearch,
  SearchOptions,
} from '../types';

/**
 * Waits for the results from the search instance to coordinate the next steps
 * in `getServerState()`.
 */
export function waitForResults(
  search: InstantSearch,
  skipRecommend: boolean = false
): Promise<SearchOptions[]> {
  const helper = search.mainHelper!;

  // Extract search parameters from the search client to use them
  // later during hydration.
  let requestParamsList: SearchOptions[];
  const client = helper.getClient();
  helper.setClient({
    ...client,
    search(queries) {
      requestParamsList = queries.map(({ params }) => params);
      return client.search(queries);
    },
  });

  search._hasSearchWidget && helper.searchOnlyWithDerivedHelpers();
  !skipRecommend && search._hasRecommendWidget && helper.recommend();

  return new Promise((resolve, reject) => {
    let searchResultsReceived = !search._hasSearchWidget;
    let recommendResultsReceived = !search._hasRecommendWidget || skipRecommend;
    // All derived helpers resolve in the same tick so we're safe only relying
    // on the first one.
    helper.derivedHelpers[0].on('result', () => {
      searchResultsReceived = true;
      if (recommendResultsReceived) {
        resolve(requestParamsList!);
      }
    });
    helper.derivedHelpers[0].on('recommend:result', () => {
      recommendResultsReceived = true;
      if (searchResultsReceived) {
        resolve(requestParamsList!);
      }
    });

    // However, we listen to errors that can happen on any derived helper because
    // any error is critical.
    helper.on('error', (error) => {
      reject(error);
    });
    search.on('error', (error) => {
      reject(error);
    });
    helper.derivedHelpers.forEach((derivedHelper) =>
      derivedHelper.on('error', (error) => {
        reject(error);
      })
    );
  });
}

/**
 * Walks the InstantSearch root index to construct the initial results.
 */
export function getInitialResults(
  rootIndex: IndexWidget,
  /**
   * Search parameters sent to the search client,
   * returned by `waitForResults()`.
   */
  requestParamsList?: SearchOptions[]
): InitialResults {
  const initialResults: InitialResults = {};

  let requestParamsIndex = 0;
  walkIndex(rootIndex, (widget) => {
    const searchResults = widget.getResults();
    const recommendResults = widget.getHelper()?.lastRecommendResults;
    if (searchResults || recommendResults) {
      const resultsCount = searchResults?._rawResults?.length || 0;
      const requestParams = resultsCount
        ? requestParamsList?.slice(
            requestParamsIndex,
            requestParamsIndex + resultsCount
          )
        : [];
      requestParamsIndex += resultsCount;
      initialResults[widget.getIndexId()] = {
        // We convert the Helper state to a plain object to pass parsable data
        // structures from server to client.
        ...(searchResults && {
          state: { ...searchResults._state },
          results: searchResults._rawResults,
        }),
        ...(recommendResults && {
          recommendResults: {
            // We have to stringify + parse because of some explicitly undefined values.
            params: JSON.parse(JSON.stringify(recommendResults._state.params)),
            results: recommendResults._rawResults,
          },
        }),
        ...(requestParams && { requestParams }),
      };
    }
  });

  if (Object.keys(initialResults).length === 0) {
    throw new Error(
      'The root index does not have any results. Make sure you have at least one widget that provides results.'
    );
  }

  return initialResults;
}
