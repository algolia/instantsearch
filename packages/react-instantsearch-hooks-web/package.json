{
  "name": "react-instantsearch-hooks-web",
  "version": "6.39.0",
  "description": "âš¡ Lightning-fast search for React, by Algolia",
  "source": "src/index.ts",
  "types": "dist/es/index.d.ts",
  "main": "dist/cjs/index.js",
  "module": "dist/es/index.js",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/es/index.js",
      "require": "./dist/cjs/index.js"
    }
  },
  "sideEffects": false,
  "license": "MIT",
  "homepage": "https://www.algolia.com/doc/guides/building-search-ui/what-is-instantsearch/react-hooks/",
  "repository": {
    "type": "git",
    "url": "https://github.com/algolia/instantsearch.js"
  },
  "author": {
    "name": "Algolia, Inc.",
    "url": "https://www.algolia.com"
  },
  "keywords": [
    "algolia",
    "components",
    "fast",
    "instantsearch",
    "react",
    "search"
  ],
  "files": [
    "README.md",
    "dist"
  ],
  "scripts": {
    "clean": "rm -rf dist",
    "watch": "yarn build:cjs --watch",
    "build": "yarn build:cjs && yarn build:es && yarn build:umd && yarn build:types",
    "build:cjs": "BABEL_ENV=cjs babel src --root-mode upward --extensions '.js,.ts,.tsx' --out-dir dist/cjs --ignore '**/__tests__/**/*','**/__mocks__/**/*' --quiet && ../../scripts/prepare-cjs.sh",
    "build:es": "BABEL_ENV=es babel src --root-mode upward --extensions '.js,.ts,.tsx' --out-dir dist/es --ignore '**/__tests__/**/*','**/__mocks__/**/*' --quiet",
    "build:umd": "BABEL_ENV=rollup rollup -c rollup.config.js",
    "build:types": "tsc -p ./tsconfig.declaration.json --outDir ./dist/es",
    "test:exports": "node ./test/module/is-es-module.mjs && node ./test/module/is-cjs-module.cjs"
  },
  "dependencies": {
    "@babel/runtime": "^7.1.2",
    "instantsearch.js": "^4.50.0",
    "react-instantsearch-hooks": "6.39.0"
  },
  "peerDependencies": {
    "algoliasearch": ">= 3.1 < 5",
    "react": ">= 16.8.0 < 19",
    "react-dom": ">= 16.8.0 < 19"
  }
}
