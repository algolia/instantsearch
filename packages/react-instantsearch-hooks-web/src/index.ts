export * from 'react-instantsearch-hooks';
export * from './widgets';
