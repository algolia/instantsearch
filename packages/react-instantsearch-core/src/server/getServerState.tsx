import {
  getInitialResults,
  waitForResults,
} from 'instantsearch.js/es/lib/server';
import { walkIndex } from 'instantsearch.js/es/lib/utils';
import React from 'react';

import { InstantSearchServerContext, InstantSearchSSRProvider } from '..';

import type {
  InstantSearchServerContextApi,
  InstantSearchServerState,
} from '..';
import type { InstantSearch, UiState } from 'instantsearch.js';
import type { ReactNode } from 'react';

type SearchRef = { current: InstantSearch | undefined };

export type RenderToString = (element: JSX.Element) => unknown;

export type GetServerStateOptions = {
  renderToString: RenderToString;
};

/**
 * Returns the InstantSearch server state from a component.
 */
export function getServerState(
  children: ReactNode,
  { renderToString }: GetServerStateOptions
): Promise<InstantSearchServerState> {
  const searchRef: SearchRef = {
    current: undefined,
  };

  const createNotifyServer = () => {
    let hasBeenNotified = false;
    const notifyServer: InstantSearchServerContextApi<
      UiState,
      UiState
    >['notifyServer'] = ({ search }) => {
      if (hasBeenNotified) {
        throw new Error(
          'getServerState should be called with a single InstantSearchSSRProvider and a single InstantSearch component.'
        );
      }

      hasBeenNotified = true;
      searchRef.current = search;
    };

    return notifyServer;
  };

  return execute({
    children,
    renderToString,
    searchRef,
    notifyServer: createNotifyServer(),
  }).then((serverState) => {
    let shouldRefetch = false;

    // <DynamicWidgets> requires another query to retrieve the dynamic widgets
    // to render.
    walkIndex(searchRef.current!.mainIndex, (index) => {
      shouldRefetch =
        shouldRefetch ||
        index
          .getWidgets()
          .some((widget) => widget.$$type === 'ais.dynamicWidgets');
    });

    if (shouldRefetch) {
      return execute({
        children: (
          <InstantSearchSSRProvider {...serverState}>
            {children}
          </InstantSearchSSRProvider>
        ),
        renderToString,
        searchRef,
        notifyServer: createNotifyServer(),
      });
    }

    return serverState;
  });
}

type ExecuteArgs = {
  children: ReactNode;
  renderToString: RenderToString;
  notifyServer: InstantSearchServerContextApi<UiState, UiState>['notifyServer'];
  searchRef: SearchRef;
};

function execute({
  children,
  renderToString,
  notifyServer,
  searchRef,
}: ExecuteArgs) {
  return Promise.resolve()
    .then(() => {
      renderToString(
        <InstantSearchServerContext.Provider value={{ notifyServer }}>
          {children}
        </InstantSearchServerContext.Provider>
      );
    })
    .then(
      () =>
        // We wait for the component to mount so that `notifyServer()` is called.
        new Promise((resolve) => setTimeout(resolve, 0))
    )
    .then(() => {
      // If `notifyServer()` is not called by then, it means that <InstantSearch>
      // wasn't within the `children`.
      // We decide to go with a strict behavior in that case; throwing. If users have
      // some routes that don't mount the <InstantSearch> component, they would need
      // to try/catch the `getServerState()` call.
      // If this behavior turns out to be too strict for many users, we can decide
      // to warn instead of throwing.
      if (!searchRef.current) {
        throw new Error(
          "Unable to retrieve InstantSearch's server state in `getServerState()`. Did you mount the <InstantSearch> component?"
        );
      }

      return waitForResults(searchRef.current);
    })
    .then(() => {
      return {
        initialResults: getInitialResults(searchRef.current!.mainIndex),
      };
    });
}
