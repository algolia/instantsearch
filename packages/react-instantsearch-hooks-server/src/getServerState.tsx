import {
  getInitialResults,
  waitForResults,
} from 'instantsearch.js/es/lib/server';
import { walkIndex } from 'instantsearch.js/es/lib/utils';
import React from 'react';
import {
  InstantSearchServerContext,
  InstantSearchSSRProvider,
} from 'react-instantsearch-hooks';

import type { InstantSearch, UiState } from 'instantsearch.js';
import type { ReactNode } from 'react';
import type {
  InstantSearchServerContextApi,
  InstantSearchServerState,
} from 'react-instantsearch-hooks';

type SearchRef = { current: InstantSearch | undefined };

export type RenderToString = (element: JSX.Element) => unknown;

export type GetServerStateOptions = {
  renderToString?: RenderToString;
};

/**
 * Returns the InstantSearch server state from a component.
 */
export function getServerState(
  children: ReactNode,
  options: GetServerStateOptions = {}
): Promise<InstantSearchServerState> {
  const searchRef: SearchRef = {
    current: undefined,
  };

  const notifyServer: InstantSearchServerContextApi<
    UiState,
    UiState
  >['notifyServer'] = ({ search }) => {
    searchRef.current = search;
  };

  return importRenderToString(options.renderToString)
    .then((renderToString) => {
      return execute({
        children,
        renderToString,
        searchRef,
        notifyServer,
      }).then((serverState) => ({ serverState, renderToString }));
    })
    .then(({ renderToString, serverState }) => {
      let shouldRefetch = false;

      // <DynamicWidgets> requires another query to retrieve the dynamic widgets
      // to render.
      walkIndex(searchRef.current!.mainIndex, (index) => {
        shouldRefetch =
          shouldRefetch ||
          index
            .getWidgets()
            .some((widget) => widget.$$type === 'ais.dynamicWidgets');
      });

      if (shouldRefetch) {
        return execute({
          children: (
            <InstantSearchSSRProvider {...serverState}>
              {children}
            </InstantSearchSSRProvider>
          ),
          renderToString,
          searchRef,
          notifyServer,
        });
      }

      return serverState;
    });
}

type ExecuteArgs = {
  children: ReactNode;
  renderToString: RenderToString;
  notifyServer: InstantSearchServerContextApi<UiState, UiState>['notifyServer'];
  searchRef: SearchRef;
};

function execute({
  children,
  renderToString,
  notifyServer,
  searchRef,
}: ExecuteArgs) {
  renderToString(
    <InstantSearchServerContext.Provider value={{ notifyServer }}>
      {children}
    </InstantSearchServerContext.Provider>
  );

  // We wait for the component to mount so that `notifyServer()` is called.
  return new Promise((resolve) => setTimeout(resolve, 0))
    .then(() => {
      // If `notifyServer()` is not called by then, it means that <InstantSearch>
      // wasn't within the `children`.
      // We decide to go with a strict behavior in that case; throwing. If users have
      // some routes that don't mount the <InstantSearch> component, they would need
      // to try/catch the `getServerState()` call.
      // If this behavior turns out to be too strict for many users, we can decide
      // to warn instead of throwing.
      if (!searchRef.current) {
        throw new Error(
          "Unable to retrieve InstantSearch's server state in `getServerState()`. Did you mount the <InstantSearch> component?"
        );
      }

      return waitForResults(searchRef.current);
    })
    .then(() => {
      return {
        initialResults: getInitialResults(searchRef.current!.mainIndex),
      };
    });
}

function importRenderToString(
  renderToString?: RenderToString
): Promise<RenderToString> {
  if (renderToString) {
    return Promise.resolve(renderToString);
  }
  // eslint-disable-next-line no-console
  console.warn(
    '[InstantSearch] `renderToString` should be passed to getServerState(<App/>, { renderToString })'
  );

  // React pre-18 doesn't use `exports` in package.json, requiring a fully resolved path
  // Thus, only one of these imports is correct
  const modules = ['react-dom/server.js', 'react-dom/server'];

  // import is an expression to make sure https://github.com/webpack/webpack/issues/13865 does not kick in
  return Promise.all(modules.map((mod) => import(mod).catch(() => {}))).then(
    (imports: unknown[]) => {
      const ReactDOMServer = imports.find(
        (mod): mod is { renderToString: RenderToString } => mod !== undefined
      );

      if (!ReactDOMServer) {
        throw new Error(
          'Could not import ReactDOMServer. You can provide it as an argument: getServerState(<Search />, { renderToString }).'
        );
      }

      return ReactDOMServer.renderToString;
    }
  );
}
