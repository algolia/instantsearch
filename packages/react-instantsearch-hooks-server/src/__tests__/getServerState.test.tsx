/**
 * @jest-environment node
 */

import {
  createMultiSearchResponse,
  createSearchClient,
  createAlgoliaSearchClient,
  createSingleSearchResponse,
  defaultRenderingContent,
} from '@instantsearch/mocks';
import React, { version as ReactVersion } from 'react';
import { renderToString } from 'react-dom/server';
import {
  InstantSearch,
  InstantSearchSSRProvider,
  Index,
  DynamicWidgets,
  version,
  useSearchBox,
} from 'react-instantsearch-hooks';
import { Hits, RefinementList } from 'react-instantsearch-hooks-web';

import { getServerState } from '../getServerState';

import type { MockSearchClient } from '@instantsearch/mocks';
import type { Hit as AlgoliaHit } from 'instantsearch.js';
import type {
  InstantSearchServerState,
  InstantSearchProps,
} from 'react-instantsearch-hooks';

function SearchBox() {
  const { query } = useSearchBox();

  return (
    <div className="ais-SearchBox">
      <form action="" className="ais-SearchBox-form" noValidate>
        <input className="ais-SearchBox-input" defaultValue={query} />
      </form>
    </div>
  );
}

function Hit({ hit }: { hit: AlgoliaHit }) {
  return <>{hit.objectID}</>;
}

type CreateTestEnvironmentProps = {
  searchClient: InstantSearchProps['searchClient'];
  initialUiState?: InstantSearchProps['initialUiState'];
};

function createTestEnvironment({
  searchClient,
  initialUiState = {
    instant_search: {
      query: 'iphone',
      refinementList: {
        brand: ['Apple'],
      },
    },
    instant_search_price_asc: {
      query: 'iphone',
      refinementList: {
        brand: ['Apple'],
      },
    },
  },
}: CreateTestEnvironmentProps) {
  function Search({ children }: { children?: React.ReactNode }) {
    return (
      <InstantSearch
        searchClient={searchClient}
        indexName="instant_search"
        initialUiState={initialUiState}
      >
        {children}
        <RefinementList attribute="brand" />
        <SearchBox />

        <h2>instant_search</h2>
        <Hits hitComponent={Hit} />

        <Index indexName="instant_search_price_asc">
          <h2>instant_search_price_asc</h2>
          <Hits hitComponent={Hit} />

          <Index indexName="instant_search_rating_desc">
            <h2>instant_search_rating_desc</h2>
            <Hits hitComponent={Hit} />
          </Index>
        </Index>

        <Index indexName="instant_search_price_desc">
          <h2>instant_search_price_desc</h2>
          <Hits hitComponent={Hit} />
        </Index>
      </InstantSearch>
    );
  }

  function App({
    serverState,
    children,
  }: {
    serverState?: InstantSearchServerState;
    children?: React.ReactNode;
  }) {
    return (
      <InstantSearchSSRProvider {...serverState}>
        <Search>{children}</Search>
      </InstantSearchSSRProvider>
    );
  }

  return {
    App,
  };
}

describe('getServerState', () => {
  test('throws without <InstantSearch> component', async () => {
    function App() {
      return null;
    }

    await expect(
      getServerState(<App />, { renderToString })
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Unable to retrieve InstantSearch's server state in \`getServerState()\`. Did you mount the <InstantSearch> component?"`
    );
  });

  test('throws with multiple <InstantSearch> components', async () => {
    function App({ serverState }: { serverState?: InstantSearchServerState }) {
      return (
        <InstantSearchSSRProvider {...serverState}>
          <InstantSearch
            searchClient={createSearchClient({})}
            indexName="index"
          />
          <InstantSearch
            searchClient={createSearchClient({})}
            indexName="index"
          />
        </InstantSearchSSRProvider>
      );
    }

    await expect(
      getServerState(<App />, { renderToString })
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"getServerState should be called with a single InstantSearchSSRProvider and a single InstantSearch component."`
    );
  });

  test('throws when the search client errors, async', async () => {
    const searchClient = createSearchClient({
      // Function is async to match the real search client.
      // a synchronous error would be caught by the derived helper, async by InstantSearch.
      // eslint-disable-next-line require-await
      search: async () => {
        throw new Error('Search client error');
      },
    });
    const { App } = createTestEnvironment({
      searchClient,
    });

    await expect(getServerState(<App />, { renderToString })).rejects.toThrow(
      /Search client error/
    );
  });

  test('throws when the search client errors, sync', async () => {
    const searchClient = createSearchClient({
      search: () => {
        throw new Error('Search client error');
      },
    });
    const { App } = createTestEnvironment({
      searchClient,
    });

    await expect(getServerState(<App />, { renderToString })).rejects.toThrow(
      /Search client error/
    );
  });

  test('adds the server user agents', async () => {
    const nextRuntime = 'nodejs';
    process.env.NEXT_RUNTIME = nextRuntime;

    const searchClient = createAlgoliaSearchClient({});
    const { App } = createTestEnvironment({ searchClient });

    await getServerState(<App />, { renderToString });

    expect(searchClient.addAlgoliaAgent).toHaveBeenCalledWith(
      `react (${ReactVersion})`
    );
    expect(searchClient.addAlgoliaAgent).toHaveBeenCalledWith(
      `react-instantsearch (${version})`
    );
    expect(searchClient.addAlgoliaAgent).toHaveBeenCalledWith(
      `react-instantsearch-hooks (${version})`
    );
    expect(searchClient.addAlgoliaAgent).toHaveBeenCalledWith(
      `react-instantsearch-server (${version})`
    );
    expect(searchClient.addAlgoliaAgent).toHaveBeenCalledWith(
      `next.js (${nextRuntime})`
    );
  });

  test('calls search with widgets search parameters', async () => {
    const searchClient = createSearchClient({});
    const { App } = createTestEnvironment({ searchClient });

    await getServerState(<App />, { renderToString });

    expect(searchClient.search).toHaveBeenCalledTimes(1);
    expect(searchClient.search).toHaveBeenCalledWith([
      {
        indexName: 'instant_search',
        params: {
          facetFilters: [['brand:Apple']],
          facets: ['brand'],
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          maxValuesPerFacet: 10,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search',
        params: {
          analytics: false,
          clickAnalytics: false,
          facets: 'brand',
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          hitsPerPage: 0,
          maxValuesPerFacet: 10,
          query: 'iphone',
          page: 0,
        },
      },
      {
        indexName: 'instant_search_price_asc',
        params: {
          facetFilters: [['brand:Apple']],
          facets: ['brand'],
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          maxValuesPerFacet: 10,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search_price_asc',
        params: {
          analytics: false,
          clickAnalytics: false,
          facets: 'brand',
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          hitsPerPage: 0,
          maxValuesPerFacet: 10,
          page: 0,
          query: 'iphone',
        },
      },
      {
        indexName: 'instant_search_rating_desc',
        params: {
          facetFilters: [['brand:Apple']],
          facets: ['brand'],
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          maxValuesPerFacet: 10,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search_rating_desc',
        params: {
          analytics: false,
          clickAnalytics: false,
          facets: 'brand',
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          hitsPerPage: 0,
          maxValuesPerFacet: 10,
          page: 0,
          query: 'iphone',
        },
      },
      {
        indexName: 'instant_search_price_desc',
        params: {
          facetFilters: [['brand:Apple']],
          facets: ['brand'],
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          maxValuesPerFacet: 10,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search_price_desc',
        params: {
          analytics: false,
          clickAnalytics: false,
          facets: 'brand',
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          hitsPerPage: 0,
          maxValuesPerFacet: 10,
          page: 0,
          query: 'iphone',
        },
      },
    ]);
  });

  test('returns initialResults', async () => {
    const searchClient = createAlgoliaSearchClient({});
    const { App } = createTestEnvironment({ searchClient });

    const serverState = await getServerState(<App />, { renderToString });

    expect(Object.keys(serverState.initialResults)).toHaveLength(4);
    expect(serverState.initialResults).toMatchSnapshot();
  });

  test('searches twice (cached) with dynamic widgets', async () => {
    const searchClient = createAlgoliaSearchClient({});
    const { App } = createTestEnvironment({ searchClient, initialUiState: {} });

    await getServerState(
      <App>
        <DynamicWidgets fallbackComponent={RefinementList} />
      </App>,
      { renderToString }
    );

    expect(searchClient.search).toHaveBeenCalledTimes(2);
    // both calls are the same, so they're cached
    expect(searchClient.search.mock.calls[0][0]).toEqual(
      searchClient.search.mock.calls[1][0]
    );
  });

  test('searches twice (cached) with dynamic widgets inside index', async () => {
    const searchClient = createAlgoliaSearchClient({});
    const { App } = createTestEnvironment({ searchClient, initialUiState: {} });

    await getServerState(
      <App>
        <Index indexName="something">
          <DynamicWidgets fallbackComponent={RefinementList} />
        </Index>
      </App>,
      { renderToString }
    );

    expect(searchClient.search).toHaveBeenCalledTimes(2);
    // both calls are the same, so they're cached
    expect(searchClient.search.mock.calls[0][0]).toEqual(
      searchClient.search.mock.calls[1][0]
    );
  });

  test('searches twice with dynamic widgets and a refinement', async () => {
    const searchClient = createAlgoliaSearchClient({
      search: jest.fn((requests) => {
        return Promise.resolve(
          createMultiSearchResponse(
            ...requests.map(() =>
              createSingleSearchResponse({
                renderingContent: defaultRenderingContent,
              })
            )
          )
        );
      }),
    });
    const { App } = createTestEnvironment({
      searchClient,
      initialUiState: {
        instant_search: {
          refinementList: {
            categories: ['refined!'],
          },
        },
      },
    });

    await getServerState(
      <App>
        <DynamicWidgets fallbackComponent={RefinementList} />
      </App>,
      { renderToString }
    );

    expect(searchClient.search).toHaveBeenCalledTimes(2);

    // first query doesn't have the fallback widget mounted yet
    expect(searchClient.search.mock.calls[0][0][0]).toEqual({
      indexName: 'instant_search',
      params: {
        facets: ['*'],
        highlightPostTag: '__/ais-highlight__',
        highlightPreTag: '__ais-highlight__',
        maxValuesPerFacet: 20,
        query: '',
        tagFilters: '',
      },
    });

    // second query does have the fallback widget mounted, and thus also the refinement
    expect(searchClient.search.mock.calls[1][0][0]).toEqual({
      indexName: 'instant_search',
      params: {
        facetFilters: [['categories:refined!']],
        facets: ['*'],
        highlightPostTag: '__/ais-highlight__',
        highlightPreTag: '__ais-highlight__',
        maxValuesPerFacet: 20,
        query: '',
        tagFilters: '',
      },
    });
  });

  test('returns HTML from server state', async () => {
    const searchClient = createSearchClient({
      search: jest.fn((requests) =>
        Promise.resolve(
          createMultiSearchResponse(
            ...requests.map(() =>
              createSingleSearchResponse({
                hits: [{ objectID: '1' }, { objectID: '2' }],
              })
            )
          )
        )
      ) as MockSearchClient['search'],
    });
    const { App } = createTestEnvironment({ searchClient });

    const serverState = await getServerState(<App />, { renderToString });
    const html = renderToString(<App serverState={serverState} />);

    expect(html).toMatchInlineSnapshot(`
      <div class="ais-RefinementList ais-RefinementList--noRefinement">
        <ul class="ais-RefinementList-list">
        </ul>
      </div>
      <div class="ais-SearchBox">
        <form action
              class="ais-SearchBox-form"
              novalidate
        >
          <input class="ais-SearchBox-input"
                 value="iphone"
          >
        </form>
      </div>
      <h2>
        instant_search
      </h2>
      <div class="ais-Hits">
        <ol class="ais-Hits-list">
          <li class="ais-Hits-item">
            1
          </li>
          <li class="ais-Hits-item">
            2
          </li>
        </ol>
      </div>
      <h2>
        instant_search_price_asc
      </h2>
      <div class="ais-Hits">
        <ol class="ais-Hits-list">
          <li class="ais-Hits-item">
            1
          </li>
          <li class="ais-Hits-item">
            2
          </li>
        </ol>
      </div>
      <h2>
        instant_search_rating_desc
      </h2>
      <div class="ais-Hits">
        <ol class="ais-Hits-list">
          <li class="ais-Hits-item">
            1
          </li>
          <li class="ais-Hits-item">
            2
          </li>
        </ol>
      </div>
      <h2>
        instant_search_price_desc
      </h2>
      <div class="ais-Hits">
        <ol class="ais-Hits-list">
          <li class="ais-Hits-item">
            1
          </li>
          <li class="ais-Hits-item">
            2
          </li>
        </ol>
      </div>
    `);
  });
});
