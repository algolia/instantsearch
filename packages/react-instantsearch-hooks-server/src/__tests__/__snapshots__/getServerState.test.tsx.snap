// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getServerState returns initialResults 1`] = `
{
  "instant_search": {
    "results": [
      {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": [],
        "hitsPerPage": 20,
        "index": "instant_search",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "clickAnalytics=true&facetFilters=%5B%5B%22brand%3AApple%22%5D%5D&facets=%5B%22brand%22%5D&highlightPostTag=__%2Fais-highlight__&highlightPreTag=__ais-highlight__&maxValuesPerFacet=10&query=iphone&tagFilters=",
        "processingTimeMS": 0,
        "query": "",
      },
      {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": [],
        "hitsPerPage": 20,
        "index": "instant_search",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "analytics=false&clickAnalytics=false&facets=brand&highlightPostTag=__%2Fais-highlight__&highlightPreTag=__ais-highlight__&hitsPerPage=0&maxValuesPerFacet=10&page=0&query=iphone",
        "processingTimeMS": 0,
        "query": "",
      },
    ],
    "state": {
      "clickAnalytics": true,
      "disjunctiveFacets": [
        "brand",
      ],
      "disjunctiveFacetsRefinements": {
        "brand": [
          "Apple",
        ],
      },
      "facets": [],
      "facetsExcludes": {},
      "facetsRefinements": {},
      "hierarchicalFacets": [],
      "hierarchicalFacetsRefinements": {},
      "highlightPostTag": "__/ais-highlight__",
      "highlightPreTag": "__ais-highlight__",
      "index": "instant_search",
      "maxValuesPerFacet": 10,
      "numericRefinements": {},
      "query": "iphone",
      "tagRefinements": [],
    },
  },
  "instant_search_price_asc": {
    "results": [
      {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": [],
        "hitsPerPage": 20,
        "index": "instant_search_price_asc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "clickAnalytics=true&facetFilters=%5B%5B%22brand%3AApple%22%5D%5D&facets=%5B%22brand%22%5D&highlightPostTag=__%2Fais-highlight__&highlightPreTag=__ais-highlight__&maxValuesPerFacet=10&query=iphone&tagFilters=",
        "processingTimeMS": 0,
        "query": "",
      },
      {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": [],
        "hitsPerPage": 20,
        "index": "instant_search_price_asc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "analytics=false&clickAnalytics=false&facets=brand&highlightPostTag=__%2Fais-highlight__&highlightPreTag=__ais-highlight__&hitsPerPage=0&maxValuesPerFacet=10&page=0&query=iphone",
        "processingTimeMS": 0,
        "query": "",
      },
    ],
    "state": {
      "disjunctiveFacets": [],
      "disjunctiveFacetsRefinements": {},
      "facets": [],
      "facetsExcludes": {},
      "facetsRefinements": {},
      "hierarchicalFacets": [],
      "hierarchicalFacetsRefinements": {},
      "highlightPostTag": "__/ais-highlight__",
      "highlightPreTag": "__ais-highlight__",
      "index": "instant_search_price_asc",
      "numericRefinements": {},
      "tagRefinements": [],
    },
  },
  "instant_search_price_desc": {
    "results": [
      {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": [],
        "hitsPerPage": 20,
        "index": "instant_search_price_desc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "clickAnalytics=true&facetFilters=%5B%5B%22brand%3AApple%22%5D%5D&facets=%5B%22brand%22%5D&highlightPostTag=__%2Fais-highlight__&highlightPreTag=__ais-highlight__&maxValuesPerFacet=10&query=iphone&tagFilters=",
        "processingTimeMS": 0,
        "query": "",
      },
      {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": [],
        "hitsPerPage": 20,
        "index": "instant_search_price_desc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "analytics=false&clickAnalytics=false&facets=brand&highlightPostTag=__%2Fais-highlight__&highlightPreTag=__ais-highlight__&hitsPerPage=0&maxValuesPerFacet=10&page=0&query=iphone",
        "processingTimeMS": 0,
        "query": "",
      },
    ],
    "state": {
      "disjunctiveFacets": [],
      "disjunctiveFacetsRefinements": {},
      "facets": [],
      "facetsExcludes": {},
      "facetsRefinements": {},
      "hierarchicalFacets": [],
      "hierarchicalFacetsRefinements": {},
      "highlightPostTag": "__/ais-highlight__",
      "highlightPreTag": "__ais-highlight__",
      "index": "instant_search_price_desc",
      "numericRefinements": {},
      "tagRefinements": [],
    },
  },
  "instant_search_rating_desc": {
    "results": [
      {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": [],
        "hitsPerPage": 20,
        "index": "instant_search_rating_desc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "clickAnalytics=true&facetFilters=%5B%5B%22brand%3AApple%22%5D%5D&facets=%5B%22brand%22%5D&highlightPostTag=__%2Fais-highlight__&highlightPreTag=__ais-highlight__&maxValuesPerFacet=10&query=iphone&tagFilters=",
        "processingTimeMS": 0,
        "query": "",
      },
      {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": [],
        "hitsPerPage": 20,
        "index": "instant_search_rating_desc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "analytics=false&clickAnalytics=false&facets=brand&highlightPostTag=__%2Fais-highlight__&highlightPreTag=__ais-highlight__&hitsPerPage=0&maxValuesPerFacet=10&page=0&query=iphone",
        "processingTimeMS": 0,
        "query": "",
      },
    ],
    "state": {
      "disjunctiveFacets": [],
      "disjunctiveFacetsRefinements": {},
      "facets": [],
      "facetsExcludes": {},
      "facetsRefinements": {},
      "hierarchicalFacets": [],
      "hierarchicalFacetsRefinements": {},
      "highlightPostTag": "__/ais-highlight__",
      "highlightPreTag": "__ais-highlight__",
      "index": "instant_search_rating_desc",
      "numericRefinements": {},
      "tagRefinements": [],
    },
  },
}
`;
